torchtune/modules/transformer.py

```diff
@@ -61,23 +103,126 @@ def forward(
+        mlp: nn.Module,
+        *,
+        attn_norm: Optional[nn.Module] = None,
+        mlp_norm: Optional[nn.Module] = None,
+        attn_scale: Optional[nn.Module] = None,
+        mlp_scale: Optional[nn.Module] = None,
+    ) -> None:
+        super().__init__()
+        assert (
+            attn.pos_embeddings is None
+        ), "Positions are not computed for encoder inputs"
+        self.attn = attn
+        self.mlp = mlp
+        self.attn_norm = attn_norm or nn.Identity()
+        self.mlp_norm = mlp_norm or nn.Identity()
+        self.attn_scale = attn_scale or nn.Identity()
+        self.mlp_scale = mlp_scale or nn.Identity()
+
+    def output_mask(self, mask: Optional[Tensor]) -> Optional[Tensor]:
+        if mask is None:
+            return None
+        if mask.dtype == torch.bool:
+            mask = ~mask
+        else:
+            mask = torch.isneginf(mask)
+        mask = torch.all(mask, dim=-1, keepdim=True)
+        return mask
+
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+    ) -> Tensor:
+        """
+        Args:
+            x (Tensor): input tensor with shape
+                [batch_size x seq_length x embed_dim]
+            mask (Optional[Tensor]): Optional boolean tensor which contains the attention mask
+                with shape [batch_size x seq_length x seq_length]. This is applied after
+                the query-key multiplication and before the softmax. A value of True in row i
+                and column j means token i attends to token j. A value of False means token i
+                does not attend to token j. If no mask is specified, a causal mask
+                is used by default. Default is None.
+            encoder_input (Optional[Tensor]): Optional second input to cross attend with x. Shape
+                [batch_size x seq_length x embed_dim] (seq_length and embed_dim may very from x)
+            encoder_mask (Optional[Tensor]):  Cross attention boolean tensor with shape
+                [batch_size x x_seq_len x encoder_seq_len]
+            input_pos (Optional[Tensor]): Optional tensor which contains the position ids
+                of each token. During training, this is used to indicate the positions
+                of each token relative to its sample when packed, shape [b x s].
+                During inference, this indicates the position of the current token.
+                If none, assume the index of the token is its position id. Default is None.
 
-        TODO:
-            - Make position of norm configurable
+        Returns:
+            Tensor: output tensor with same shape as input
+                [batch_size x seq_length x embed_dim]
         """
+        # Skip cross attention when no secondary input
+        if encoder_input is None:
```

Comment:
I talked with @pbontrager about this a bit already. The standard implementation of a transformer decoder layer with cross-attention contains both self-attention and cross-attention in a single layer. If encoder inputs are not passed one would simply bypass the cross-attention but still hit the self-attention. But our implementation splits self-attention and cross-attention into their own layers. In that case the analogous thing to do would be just make this entire layer a no-op. 

But I do think your question is a valid one -- like should we ever actually hit this without encoder hidden states?

Reply:
The simplest case would be during inference where you have a multimodal model and the user hasn't provided an image. But this could happen during finetuning too if not every data sample had images etc.